### PR TITLE
Adds `build_histogram(Equalization(),...)`

### PR DESCRIPTION
**Pull request by zygmuntszpak**
Removed the dependence on `searchsortedlast` from `build_histogram` to improve the performance of histogram construction. Added a fresh implementation of histogram equalization which  does not require `_histeq_pixel_rescale`.

### CODE REVIEW

File: src/exposure.jl [hunk 1]
```diff
@@ -482,6 +488,154 @@ end
...
+    edges, histogram = build_histogram(img, nbins, minval, maxval)
+    lb = first(axes(histogram,1))
+    ub = last(axes(histogram,1))
+    N = length(img)
+    cdf = cumsum(histogram[lb:ub]/N)
+    transform_density!(img, edges, histogram, cdf)
+end
+
+function transform_density!(img::AbstractArray,edges::AbstractArray, histogram::AbstractArray, cdf::AbstractArray)
+    o = Base.Order.Forward
```
Comment (zygmuntszpak) [inline src/exposure.jl]: Forgot to remove this.

File: src/exposure.jl [hunk 2]
```diff
@@ -482,6 +488,154 @@ end
...
+# References
+1. R. C. Gonzalez and R. E. Woods. *Digital Image Processing (3rd Edition)*.  Upper Saddle River, NJ, USA: Prentice-Hall,  2006.
+
+See also: [histmatch](@ref),[clahe](@ref), [build_histogram](@ref) and  [adjust_gamma](@ref).
+
+"""
+function adjust_histogram(operation::Equalization, img::AbstractArray{T}, nbins::Integer, minval::RealLike = 0, maxval::RealLike = 1) where {T<:AbstractRGB}
+    yiq = convert.(YIQ, img)
+    yiq_view = channelview(yiq)
+    yeq = adjust_histogram!(Equalization(),view(yiq_view,1,:,:),nbins,minval,maxval)
```
Comment (zygmuntszpak) [inline src/exposure.jl]: Don't need to declare `yeq` since `adjust_histogram!` modifies its argument. I forgot to remove the `yeq`  after refactoring my initial implementation.

File: src/exposure.jl [hunk 3]
```diff
@@ -318,30 +318,36 @@ function build_histogram(img::AbstractArray, nbins::Integer=200)
...
+function build_histogram(img::AbstractArray{T}, edges::AbstractRange) where {T<:AbstractRGB}
+    build_histogram(Gray.(img),edges)
+end
+
 function build_histogram(img::AbstractArray, edges::AbstractRange)
+    Base.has_offset_axes(edges) && throw(ArgumentError("edges must be indexed starting with 1"))
     lb = first(axes(edges,1))-1
     ub = last(axes(edges,1))
+    first_edge = first(edges)
+    step_size = edges.step.hi
```
Comment (timholy) [inline src/exposure.jl]: This seems to assume a `TwicePrecision` element. But I can construct an AbstractRange as `0:11` or `StepRangeLen(0.0, 0.1, 11)`.

File: src/exposure.jl [hunk 4]
```diff
@@ -482,6 +488,154 @@ end
...
+
+# References
+1. R. C. Gonzalez and R. E. Woods. *Digital Image Processing (3rd Edition)*.  Upper Saddle River, NJ, USA: Prentice-Hall,  2006.
+
+See also: [histmatch](@ref),[clahe](@ref), [build_histogram](@ref) and  [adjust_gamma](@ref).
+
+"""
+function adjust_histogram(operation::Equalization, img::AbstractArray{T}, nbins::Integer, minval::RealLike = 0, maxval::RealLike = 1) where {T<:AbstractRGB}
+    yiq = convert.(YIQ, img)
+    yiq_view = channelview(yiq)
```
Comment (timholy) [inline src/exposure.jl]: Better would be something like `y = map(c->YIQ(c).y, img)` (untested).

File: src/exposure.jl [hunk 5]
```diff
@@ -482,6 +488,154 @@ end
...
+imshow(imgeq)
+```
+
+# References
+1. R. C. Gonzalez and R. E. Woods. *Digital Image Processing (3rd Edition)*.  Upper Saddle River, NJ, USA: Prentice-Hall,  2006.
+
+See also: [histmatch](@ref),[clahe](@ref), [build_histogram](@ref) and  [adjust_gamma](@ref).
+
+"""
+function adjust_histogram(operation::Equalization, img::AbstractArray{T}, nbins::Integer, minval::RealLike = 0, maxval::RealLike = 1) where {T<:AbstractRGB}
```
Comment (timholy) [inline src/exposure.jl]: Does it have to be `AbstractRGB` or does any `Color3` suffice?

File: src/exposure.jl [hunk 6]
```diff
@@ -318,30 +318,36 @@ function build_histogram(img::AbstractArray, nbins::Integer=200)
     build_histogram(img, nbins, minfinite(img), maxfinite(img))
 end
 
+function build_histogram(img::AbstractArray{T}, edges::AbstractRange) where {T<:AbstractRGB}
+    build_histogram(Gray.(img),edges)
+end
+
 function build_histogram(img::AbstractArray, edges::AbstractRange)
+    Base.has_offset_axes(edges) && throw(ArgumentError("edges must be indexed starting with 1"))
```
Comment (timholy) [inline src/exposure.jl]: Seems generic so not sure this is necessary (but I have no objection to including it, either).

File: src/exposure.jl [hunk 7]
```diff
@@ -482,6 +488,154 @@ end
...
+end
+
+function transform_density!(img::AbstractArray,edges::AbstractArray, histogram::AbstractArray, cdf::AbstractArray)
+    o = Base.Order.Forward
+    first_edge = first(edges)
+    step_size = edges.step.hi
+    map!(img,img) do val
+        if val >= edges[end]
+            newval = cdf[end]
+        else
```
Comment (timholy) [inline src/exposure.jl]: do you need to check `val < first(edges)` here too?

File: test/exposure.jl [hunk 8]
```diff
@@ -77,7 +77,8 @@ eye(m,n) = Matrix{Float64}(I,m,n)
                 @test collect(counts) == [128; 16; 16; 16; 80]
                 @test axes(counts) == (0:length(edges),)
             else
-                edges, counts  = build_histogram(0:1/255:1,4,128/255,192/255)
+                img = collect(0:1/255:1)
+                edges, counts  = build_histogram(T.(img),4,128/255,192/255)
```
Comment (timholy) [inline test/exposure.jl]: Just curious, what's the reason for this change?

File: src/exposure.jl [hunk 9]
```diff
@@ -482,6 +488,154 @@ end
...
+struct Equalization <: AbstractHistogramOperation end
+
+"""
+```
+hist_equalised_img = adjust_histogram(Equalization(),img, nbins)
+hist_equalised_img = adjust_histogram(Equalization(),img, nbins, minval, maxval)
+```
+
+Returns a histogram equalised image with a granularity of approximately `nbins`
+number of bins.
```
Comment (timholy) [inline src/exposure.jl]: Can you remind me the reason for the "approximately"?

File: src/exposure.jl (hunk 6 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: It's not quite generic yet. The manner in which I calculate the index currently still assumes that edges start from 1.

File: src/exposure.jl (hunk 5 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: `Color3` is what we want. I occasionally have trouble determining what the most abstract types are---sorry.

File: src/exposure.jl (hunk 4 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: I actually need not only the `y` value, but also the `i` and `q` values since I need to reconstruct the original colour image. That's why I convert to `YIQ` and pass the `Y` as a view to be equalised.

File: src/exposure.jl (hunk 7 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: Yes I do :D.

File: test/exposure.jl (hunk 8 above)
Comment (zygmuntszpak) [reply test/exposure.jl]: I wanted to convert the `img` to all the different types `T` that we loop over in the outer loop. That was my original intention but I had forgotten to actually convert the `img` to type `T` in the previous commit.

File: src/exposure.jl (hunk 9 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: This is a copy/paste mistake. I copied the description from `imhist`. I recently added the word _approximately_ to `imhist`  because `imhist` doesn't guarantee that the number of bins that the user specifies actually are utilised in the construction of the histogram. Thanks for spotting this---I've fixed it in the latest version.

File: src/exposure.jl (hunk 3 above)
Comment (zygmuntszpak) [reply src/exposure.jl]: Thanks. I think I've handled the general case in the new version of the code.

File: src/exposure.jl (hunk 4 above)
Comment (timholy) [reply src/exposure.jl]: Good. But wouldn't it still be better to access the fields you want directly rather than using `channelview`? To me the meaning of `A[i,j].y` seems much more evident than `A[1,i,j]`.

**Comment by timholy** [comment]
No problem about the new branch, I know how it goes. I will close this but note some responses above.
